### PR TITLE
*:Unify integer PK and handle phase1

### PR DESCRIPTION
**Pull request by coocood**
In this phase, when PK is Integer type, we use PK value as handle value, and we don't set PK column.
But we still create an index for PK, because handle range scan has not been supported.

### CODE REVIEW

File: ddl/ddl.go [hunk 1]
```diff
@@ -424,6 +424,17 @@ func (d *ddl) buildTableInfo(tableName model.CIStr, cols []*column.Col, constrai
 		tbInfo.Columns = append(tbInfo.Columns, &v.ColumnInfo)
 	}
 	for _, constr := range constraints {
+		if constr.Tp == coldef.ConstrPrimaryKey {
+			if len(constr.Keys) == 1 {
+				key := constr.Keys[0]
+				col := column.FindCol(cols, key.ColumnName)
+				switch col.Tp {
+				case mysql.TypeLong, mysql.TypeLonglong:
+					tbInfo.PKIsHandle = true
```
Comment (shenli) [inline ddl/ddl.go]: How about use auto_increment column as pk?

Comment (coocood) [inline ddl/ddl.go]: auto_increment column may not be PK.

File: stmt/stmts/insert.go [hunk 2]
```diff
@@ -368,9 +358,36 @@ func getOnDuplicateUpdateColumns(assignList []*expression.Assignment, t table.Ta
...
+					}
+					row[i] = recordID
+					// Notes: incompatible with mysql
+					// MySQL will set last insert id to the first row, as follows:
+					// `t(id int AUTO_INCREMENT, c1 int, PRIMARY KEY (id))`
+					// `insert t (c1) values(1),(2),(3);`
+					// Last insert id will be 1, not 3.
+					variable.GetSessionVars(ctx).SetLastInsertID(uint64(recordID))
+				} else {
+					// Null primary key, will be checked later.
```
Comment (shenli) [inline stmt/stmts/insert.go]: Throw error?

Comment (coocood) [inline stmt/stmts/insert.go]: The following procedure will check not null, error will be thrown there.

File: stmt/stmts/insert.go [hunk 3]
```diff
@@ -382,15 +399,18 @@ func (s *InsertValues) initDefaultValues(ctx context.Context, t table.Table, row
 		}
 
 		if mysql.HasAutoIncrementFlag(c.Flag) {
-			if recordID, err = t.AllocAutoID(); err != nil {
-				return 0, errors.Trace(err)
+			// The auto_increment column is not primary key, don't set
```
Comment (shenli) [inline stmt/stmts/insert.go]: Why add this comment? I do not see the relevance between the comment and the changes.

Comment (coocood) [inline stmt/stmts/insert.go]: I move `SetLastInserID` to right after `AllocAutoID` is called, this comment explain the difference between above code.

Comment (coocood) [inline stmt/stmts/insert.go]: The old code assume HasAutoIncrementFlag implies its primary key.

File: table/tables/tables.go [hunk 4]
```diff
@@ -104,7 +103,6 @@ func NewTable(tableID int64, tableName string, cols []*column.Col, alloc autoid.
 		indexPrefix:  string(genTableIndexPrefix(tableID)),
 		alloc:        alloc,
 		Columns:      cols,
-		state:        model.StatePublic,
```
Comment (shenli) [inline table/tables/tables.go]: Why remove this?

Comment (coocood) [inline table/tables/tables.go]: Because it's included in `meta`.

File: table/tables/tables.go [hunk 5]
```diff
@@ -406,11 +389,26 @@ func (t *Table) rebuildIndices(rm kv.RetrieverMutator, h int64, touched map[int]
...
+	var hasRecordID bool
+	for _, col := range t.Cols() {
+		if mysql.HasPriKeyFlag(col.Flag) && t.meta.PKIsHandle {
+			// The value must have been filled with int64 or uint64 value, or it will not pass null check.
+			switch x := r[col.Offset].(type) {
+			case int64:
+				recordID = x
+			case uint64:
+				recordID = int64(x)
+			case int:
```
Comment (shenli) [inline table/tables/tables.go]: Why not use types.ToInt64() here?

Comment (coocood) [inline table/tables/tables.go]: For performance, as we are sure it is already a int type.

Comment (shenli) [inline table/tables/tables.go]: Should consider uint or int16/uint16?

Comment (coocood) [inline table/tables/tables.go]: No, we parse and converts to int64 and uint64 for any types, int may used in tests, there is no other integer types used.

Comment (shenli) [inline table/tables/tables.go]: How about run prepared statement with binary protocol?

Comment (coocood) [inline table/tables/tables.go]: The value has been cast to the column type, so it is int64 or uint64, but I should use `types.ToInt64` for simplicity, just need to optimize it a little bit.

File: stmt/stmts/insert.go (hunk 2 above)
Comment (shenli) [inline stmt/stmts/insert.go]: So why not remove this else?

File: table/tables/tables.go (hunk 4 above)
Comment (shenli) [inline table/tables/tables.go]: I find t.state is set here: https://github.com/pingcap/tidb/blob/master/table/tables/tables.go#L93

But it is not related with meta. So what is meta mean?

File: table/tables/tables.go [hunk 6]
```diff
@@ -406,11 +389,19 @@ func (t *Table) rebuildIndices(rm kv.RetrieverMutator, h int64, touched map[int]
...
-func (t *Table) AddRecord(ctx context.Context, r []interface{}, h int64) (recordID int64, err error) {
-	// Already have recordID
-	if h != 0 {
-		recordID = int64(h)
-	} else {
+func (t *Table) AddRecord(ctx context.Context, r []interface{}) (recordID int64, err error) {
+	var hasRecordID bool
+	for _, col := range t.Cols() {
+		if mysql.HasPriKeyFlag(col.Flag) && t.meta.PKIsHandle {
+			recordID, err = types.ToInt64(r[col.Offset])
```
Comment (shenli) [inline table/tables/tables.go]: You have already converted r[col.Offset] value to int64 in initDefaultValues(). So why do it here again?

Comment (coocood) [inline table/tables/tables.go]: Here we need to return an `int64` but `r[col.Offset]` is an interface{}

File: ddl/ddl.go [hunk 7]
```diff
@@ -424,6 +424,17 @@ func (d *ddl) buildTableInfo(tableName model.CIStr, cols []*column.Col, constrai
 		tbInfo.Columns = append(tbInfo.Columns, &v.ColumnInfo)
 	}
 	for _, constr := range constraints {
+		if constr.Tp == coldef.ConstrPrimaryKey {
+			if len(constr.Keys) == 1 {
+				key := constr.Keys[0]
+				col := column.FindCol(cols, key.ColumnName)
+				switch col.Tp {
```
Comment (disksing) [inline ddl/ddl.go]: No need to check `col != nil`?

File: util/types/convert.go [hunk 8]
```diff
@@ -559,6 +559,16 @@ func ToUint64(value interface{}) (uint64, error) {
 
 // ToInt64 converts an interface to an int64.
 func ToInt64(value interface{}) (int64, error) {
+	switch x := value.(type) {
```
Comment (qiuyesuifeng) [inline util/types/convert.go]: Why add some logic here?

File: stmt/stmts/insert.go [hunk 9]
```diff
@@ -368,9 +358,36 @@ func getOnDuplicateUpdateColumns(assignList []*expression.Assignment, t table.Ta
...
+func (s *InsertValues) initDefaultValues(ctx context.Context, t table.Table, row []interface{}, marked map[int]struct{}) error {
 	var defaultValueCols []*column.Col
 	for i, c := range t.Cols() {
+		if mysql.HasPriKeyFlag(c.Flag) && t.Meta().PKIsHandle {
+			if row[i] != nil {
+				recordID, err := types.ToInt64(row[i])
+				if err != nil {
+					return errors.Trace(err)
+				}
+				row[i] = recordID
```
Comment (disksing) [inline stmt/stmts/insert.go]: It seems that above lines inside `if row[i] != nil` can be removed, as the type will be checked in `CastValues()`.

Comment (disksing) [inline stmt/stmts/insert.go]: And If that is true, seems that we can remove the whole `if PKIsHandle {}` and add `if PKIsHandle {SetLastInsertID}` logic in `mysql.HasAutoIncrementFlag()` below.

File: util/types/convert.go (hunk 8 above)
Comment (coocood) [inline util/types/convert.go]: To improve performance.

File: ddl/ddl.go (hunk 7 above)
Comment (coocood) [inline ddl/ddl.go]: I should check it.

File: table/tables/tables.go (hunk 4 above)
Comment (coocood) [inline table/tables/tables.go]: `meta` is the table schema information.
The table instance should keep a reference to it.

If we have the reference, we don't need to copy the field value in meta.

**Comment by siddontang** [comment]
@zimulala 
Does the inspector have the test for checking primary key? the primary key can't exist in index now.

**Comment by zimulala** [comment]
@siddontang 
no. 
I think it can be added after coocood finishes to unify pk.

**Comment by ngaut** [comment]
LGTM

**Comment by ngaut** [comment]
@siddontang PTAL

File: stmt/stmts/insert.go [hunk 10]
```diff
@@ -382,15 +371,24 @@ func (s *InsertValues) initDefaultValues(ctx context.Context, t table.Table, row
...
 
 		if mysql.HasAutoIncrementFlag(c.Flag) {
-			if recordID, err = t.AllocAutoID(); err != nil {
-				return 0, errors.Trace(err)
+			recordID, err := t.AllocAutoID()
+			if err != nil {
+				return errors.Trace(err)
 			}
 			row[i] = recordID
+			if mysql.HasPriKeyFlag(c.Flag) && t.Meta().PKIsHandle {
```
Comment (siddontang) [inline stmt/stmts/insert.go]: may places use this condition check, can extract to a function.

File: util/types/convert.go (hunk 8 above)
Comment (siddontang) [inline util/types/convert.go]: really need to do it here now?

**Comment by siddontang** [comment]
LGTM

**Pull request merged**